### PR TITLE
vhost_user: fix unsoundness in VhostUserMsgHeader

### PR DESCRIPTION
**Pull request by alyssais**
### Summary of the PR

Since `VhostUserMsgHeader` implements `ByteValued`, it is supposed to be safe to construct from any correctly-sized arbitrary byte array. But that means we can do this:

```rust
let bytes = b"\xFF\xFF\xFF\xFF\x00\x00\x00\x00\x00\x00\x00\x00";
let header = VhostUserMsgHeader::<MasterReq>::from_slice(bytes).unwrap();
header.get_code()
```

constructing an invalid `MasterReq`, using only functions that are marked as safe.  Constructing an invalid enum value is undefined behavior in Rust, so this API is unsound.  This wasn't considered by the safety comment in `VhostUserMsgHeader::get_code`, which only considered the safety of requests that were valid enum variants.

If the vhost-user frontend process sends a message that the backend doesn't recognise, that's exactly what will happen, so the UB can be triggered from an external process (but a trusted one).

To fix this, we need to check whether the value is valid _before_ converting it.  `Req::is_valid` is changed to be a non-instance method, so it can be called before constructing the `Req`.
`VhostUserMsgHeader::get_code` is changed to return a `Result`, to accomodate the case where the request number is not a valid value for `R`.

Since this unsoundness is exposed to other processes, I will apply for an informational (not vulnerability) RUSTSEC advisory for this issue.

Signed-off-by: Alyssa Ross <alyssa.ross@unikie.com>

### Requirements

Before submitting your PR, please make sure you addressed the following
requirements:

- [x] All commits in this PR are signed (with `git commit -s`), and the commit
  message has max 60 characters for the summary and max 75 characters for each
  description line.
- [x] All added/changed functionality has a corresponding unit/integration
  test.
- [x] All added/changed public-facing functionality has entries in the "Upcoming 
  Release" section of CHANGELOG.md (if no such section exists, please create one).
- [x] Any newly added `unsafe` code is properly documented.

### CODE REVIEW

File: crates/vhost/src/vhost_user/message.rs [hunk 1]
```diff
@@ -981,38 +990,32 @@ mod tests {
...
     #[test]
     fn check_slave_request_code() {
-        let code = SlaveReq::NOOP;
-        assert!(!code.is_valid());
-        let code = SlaveReq::MAX_CMD;
-        assert!(!code.is_valid());
-        assert!(code > SlaveReq::NOOP);
+        assert!(!MasterReq::is_valid(MasterReq::NOOP as _));
+        assert!(!MasterReq::is_valid(MasterReq::MAX_CMD as _));
+        assert!(MasterReq::MAX_CMD > MasterReq::NOOP);
```
Comment (stefano-garzarella) [inline crates/vhost/src/vhost_user/message.rs:1006]: In this code block, we should use `SlaveReq::` methods/values

File: crates/vhost/src/vhost_user/slave_req_handler.rs [hunk 2]
```diff
@@ -609,11 +609,10 @@ impl<S: VhostUserSlaveReqHandler> SlaveReqHandler<S> {
...
             return Err(Error::InvalidMessage);
         }
-        let flags: VhostUserConfigFlags;
-        match VhostUserConfigFlags::from_bits(msg.flags) {
-            Some(val) => flags = val,
+        let flags: VhostUserConfigFlags = match VhostUserConfigFlags::from_bits(msg.flags) {
+            Some(val) => val,
             None => return Err(Error::InvalidMessage),
-        }
+        };
```
Comment (stefano-garzarella) [inline crates/vhost/src/vhost_user/slave_req_handler.rs:615]: This change seems unrelated (I prefer your cleanup), maybe you can do it in a separate commit.

**Review by stefano-garzarella — CHANGES_REQUESTED** [review]
I left a few small comments, but overall it looks great!

File: crates/vhost/src/vhost_user/slave_req_handler.rs (hunk 2 above)
Comment (lauralt) [reply crates/vhost/src/vhost_user/slave_req_handler.rs:615]: I think you can use `ok_or` here.

Comment (alyssais) [reply crates/vhost/src/vhost_user/slave_req_handler.rs:615]: I changed it because I saw CI failed due to clippy, but I think that may actually have been for a different reason, so I can probably drop this / make a separate PR to make clippy happy.

Comment (stefano-garzarella) [reply crates/vhost/src/vhost_user/slave_req_handler.rs:615]: Sorry, I misspoke!
The fix is fine, especially if there is a new clippy warning (I guess with a new version of Rust), I was just asking to do it in a separate commit, even within this PR :-)

File: crates/vhost/src/vhost_user/message.rs (hunk 1 above)
Comment (alyssais) [reply crates/vhost/src/vhost_user/message.rs]: Fixed.

File: crates/vhost/src/vhost_user/slave_req_handler.rs (hunk 2 above)
Comment (alyssais) [reply crates/vhost/src/vhost_user/slave_req_handler.rs]: Done. :)

**Review by stefano-garzarella — APPROVED** [review]

**Review by sboeuf — APPROVED** [review]
